### PR TITLE
internal() source running in dedicated thread

### PR DESCRIPTION
**Pull request by Kokan**
The `internal()` source currently runs in the main thread, while the main thread main goal is to orchestrate the work and not doing it.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

File: lib/afinter.c [hunk 1]
```diff
@@ -122,16 +123,71 @@ afinter_source_post(gpointer s)
       log_source_post(&self->super, msg);
     }
   afinter_source_update_watches(self);
+  main_loop_worker_invoke_batch_callbacks();
```
Comment (gaborznagy) [inline lib/afinter.c:126]: I've ran through the patches and they look good to me, but I wonder, why do we need to call `main_loop_worker_invoke_batch_callbacks`?

Comment (Kokan) [reply lib/afinter.c:126]: That's a great question. It is not something that I know should be there (and possible cannot explain in 100% accuracy).

This new thread is a worker thread (`main_loop_create_worker_thread`), it processes messages from source till destination queue.
The current implementation of queue fifo has a fast and slow path, depending on if it was called from a worker thread or a simple thread.
When a simple thread is used it stuck with the slow path, otherwise from worker thread it can use the fast path with per thread queue.

The fast path at some point shall merge the per thread queues, the current solution that fast path registers a call back doing that job later. A worker thread has a framework to deal with these callbacks, but have no mean of calling it automatically (as it should be executed in the worker thread when it has time, but regularly). 

If you remove this call two things going to happen:
1. An assert guarding the worker thread shutdown with checking for pending jobs.
2. The messages actually won't be processed by destination (in case of logqueue fifio) as they are waiting for the merger from per thread queue to the queue a destination works from.

This is my current understanding here.

**Comment by kira-syslogng** [comment]
Build FAILURE

File: lib/afinter.c (hunk 1 above)
Comment (Kokan) [reply lib/afinter.c]: Note to this, in the end I'v disabled fast path for this source. As the fast path currently entangled with the number of io worker thread, thread id assignment. The number of thread queue set based on the number of io worker. IMHO this is not right, but a separate issue instead of opening the scope I do what every other threaded src do currently and use slow path. (https://github.com/syslog-ng/syslog-ng/pull/3428/commits/e3c016ee6fbdd0cb3096af3f5d5577b3382e90fa)

It leads to crash otherwise, as the logqueue fifo assert on its own expected number of queue and the current thread value.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/afinter.c (hunk 1 above)
Comment (gaborznagy) [reply lib/afinter.c]: Thanks for the explanation. I saw that there is only one place where `main_loop_worker_register_batch_callback` is called, and it was `logqueue-fifo.c`. I've sensed a rabbit hole, and you proved it. :)
It's an interesting topic though.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by Kokan** [comment]
Only rebase to follow up macos ci fix on master, no code change since last push.

File: lib/afinter.c [hunk 2]
```diff
@@ -124,14 +125,70 @@ afinter_source_post(gpointer s)
...
+  self->free_to_send = TRUE;
+  g_static_mutex_unlock(&internal_msg_lock);
+
+  afinter_source_start_watches(self);
+  afinter_source_update_watches(self);
+
+  iv_main();
+
+  iv_event_unregister(&self->exit);
+  iv_event_unregister(&self->post);
```
Comment (MrAnno) [inline lib/afinter.c:154]: I have a crash because of a possible race, where an internal message triggers the post event, but the it has previously been unregistered.

I'm investigating this scenario.

**Review by MrAnno — CHANGES_REQUESTED** [review]

File: lib/cfg.c [hunk 3]
```diff
@@ -368,7 +368,9 @@ cfg_init(GlobalConfig *cfg)
   if (!cfg_tree_start(&cfg->tree))
     return FALSE;
 
-  g_assert(cfg_tree_on_inited(&cfg->tree));
+  if (!cfg_tree_on_inited(&cfg->tree))
+    return FALSE;
```
Comment (MrAnno) [inline lib/cfg.c:372]: Please keep the assertion here. It is intentional, see https://github.com/syslog-ng/syslog-ng/pull/3176#issuecomment-638849597.

TLDR: A half-initialized pipeline turned out to be really hard to deinitialize correctly when dedicated source/destination threads are spawned (because we would have to wait for workers to stop and guarantee some internal task/timer/fdwatch ordering in ivykis during this action).

The best would be to extract these 2 lines or just add a comment.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/cfg.c (hunk 3 above)
Comment (Kokan) [reply lib/cfg.c:372]: Thanks, I add your comment, and a link to the original comment, those should be a nice point to unroll the shenanigans of this assert.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]
Thanks!

**Comment by Kokan** [comment]
Change: the `on_config_init` was initialised in source driver init function, instead of source driver new. (Thanks for @gaborznagy noticing it.)

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by gaborznagy — APPROVED** [review]
Thanks for the changes!

**Review by MrAnno — APPROVED** [review]

**Comment by gaborznagy** [comment]
LGTM has some internal issues around APT, I'm merging this anyway (previously it was green).
```
Could not connect to archive.ubuntu.com:80 (91.189.88.142), connection timed out Could not connect to archive.ubuntu.com:80 (91.189.88.152), connection timed out
```
